### PR TITLE
use Calcites.getColumnTypeForRelDataType for SQL CAST operator conversion

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This PR modifies the SQL `CAST` operator conversion to use `Calcites.getColumnTypeForRelDataType` to convert calcite types to native Druid types instead of using its own custom `SqlTypeName ` to `ExprType` mapping to make it more consistent with other SQL to Druid type conversions done for most other operators. This allows it to handle some additional casts which were not previously supported, such as those with ARRAY types. The test added to `CalciteMultiValueStringQueryTest` was not able to be planned prior to the changes to the `CAST` operator since during planning explicit casts would be added which would then fail since the mapping table was missing arrays.

Switching `CAST` to use `Calcites.getColumnTypeForRelDataType` also required making a small adjustment to this method to be able to handle `SqlTypeName.DAY_INTERVAL_TYPES` and `SqlTypeName.YEAR_INTERVAL_TYPES` as a Druid `LONG` type. This also allows some additional expressions to be planned that were not previously supported, shown in the changes to `ExpressionsTest`, `GreatestExpressionTest`, and `LeastExpressionTest`.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by abhishekagarwal87 — APPROVED** [review]

File: sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java [hunk 1]
```diff
@@ -1779,8 +1778,7 @@ public void testTimeMinusYearMonthInterval()
             DruidExpression.functionCall("timestamp_shift"),
             ImmutableList.of(
                 DruidExpression.ofColumn(ColumnType.LONG, "t"),
-                // RexNode type "interval year to month" is not reported as ColumnType.STRING
-                DruidExpression.ofLiteral(null, DruidExpression.stringLiteral("P13M")),
+                DruidExpression.ofLiteral(ColumnType.LONG, DruidExpression.stringLiteral("P13M")),
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:1781]: Long type for string literal? Seems not right.

Comment (clintropolis) [reply sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:1781]: hmm, so this was a result of merging the rules from cast into `Calcites.getColumnTypeForRelDataType` which had interval types as LONG, but agree it does look a bit strange so will fix to report these interval types as strings instead and see what that breaks

Comment (clintropolis) [reply sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:1781]: ah, i see why it was treating the cast as LONG typed...

`timestampDiff` test for example starts spitting out explicit casts which are not necessary so I think i need to add a bit extra logic to cast to let intervals cast to longs without a cast ```ExpressionVirtualColumn{name='v0', expression='div(CAST(("__time" - 915148800000), 'LONG'),86400000)', outputType=LONG}```

Comment (clintropolis) [reply sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java]: ok, updated to handle intervals as STRINGS so they match the argument literals in plans, but also handle them as longs for casting (and also reduction). I suppose instead I could just drop intervals from being handled by `Calcites.getColumnTypeForRelDataType` like it was before, but string seems the most correct i think?

**Review by gianm — APPROVED** [review]
LGTM after the latest changes

**Pull request merged**